### ISSUE TITLE
Bitcode support for iOS frameworks

### ISSUE THREAD

**Issue by anupcowkur** [JetBrains/kotlin-native#1202]
Hi. Is there a way to generate an iOS framework that has bitcode enabled from the gradle plugin? 
Currently `./gradlew build` generates iOS frameworks for simulator and device but they don't seem to be bitcode enabled.

**Comment by olonho** [comment]
Not yet. Why do you need that functionality?

**Comment by anupcowkur** [comment]
We have a large iOS project that uses bitcode. Xcode requires that if you enable bitcode, all your framework dependencies must also be built with bitcode.

**Comment by siuying** [comment]
On watchos that’s required. It is also likely in future this would be expanded to iOS.

**Comment by anupcowkur** [comment]
Bitcode is required to be enabled only in WatchOS but if you have optionally enabled bitcode in your regular app, it is required that [all frameworks that are included are also bitcode enabled](https://developer.apple.com/library/content/documentation/IDEs/Conceptual/AppDistributionGuide/AppThinning/AppThinning.html#//apple_ref/doc/uid/TP40012582-CH35-SW2).

**Comment by olonho** [comment]
Well, currently we cannot provide bitcode compatible with what Apple needs, so maybe just avoid using bitcode in the meantime.

**Comment by kylejbrock** [comment]
Is it a technical limitation that would live past 1.0?

**Comment by oleksandr-yefremov** [comment]
> Bitcode is required to be enabled only in WatchOS

This part of statement is not correct. Both tvOS and watchOS **require** bitcode enabled now.

See https://developer.apple.com/tvos/submit/
> When compiling your tvOS app and any third-party libraries, Bitcode must be enabled


So to summarize: without bitcode one cannot develop any tvOS and watchOS frameworks/apps. 
In theory one should be able to build iOS framework/app without bitcode, but in reality there are 2 issues at least:
1. If you are building an application, then you won't be able to use closed source dependencies (prebuilt frameworks), because they are (almost) all come with bitcode enabled. You can only use open source dependencies and build them without bitcode. If an example helps, Google IMA for iOS (most widely used video ad management framework) is a prebuilt framework and comes with bitcode.
2. If you are building a framework, point 1 also applies here. Plus, you have to ship 2 different versions with/without bitcode to avoid blocking people who use/don't use bitcode in their apps. 

I completely understand that enabling bitcode is a complex feature, partially because Apple does whatever they want with LLVM. [Same issue discussion in rust](https://github.com/rust-lang/rust/issues/35968) is open for 1,5 years already and with no ETA. However, IMO, K/N is in better position than Rust, because for Rust iOS support is a low-prio "good to have" feature, whereas K/N is devoted to client side from the beginning. My team is investigating now, whether we could ditch our 2 codebases on Java for Android/AndroidTV and Swift on iOS/tvOS to build one single K/N code and test base for all platforms. Everything looks too good to be true 😄  Except if there is no ETA for bitcode (even in 6 months from now), we won't be able to ship it to customers who use bitcode for a long time now and the whole idea falls apart. K/N FTW! 🏎

**Comment by mouse4d** [comment]
I totally agree with @oleksandr-yefremov! We start a new SKD project that is 90% business logic, so we're starting it in Kotlin, with the expectations that it will advance quickly and it will support Bitcode and Dynamic Library in the next time (6 months is good enough). Will be a real blow if it doesn't, but we're willing to take the risk.

**Comment by alastaircoote** [comment]
It appears that Rust has recently added support for outputting Bitcode: https://github.com/rust-lang/rust/pull/48896, in case that work provides any useful pointers for Kotlin Native.

**Comment by AxelDelmas** [comment]
Is there any news on this topic? We're also considering K/N but support for iOS bitcode is also a blocking point for us. Thanks

**Comment by ankushg** [comment]
Created a ticket on YouTrack for this since it seems like things are more frequently responded to by Jetbrains over there: https://youtrack.jetbrains.com/issue/KT-27147

**Comment by simonbengtsson** [comment]
Does this mean bitcode is supported now? 😲

https://github.com/JetBrains/kotlin-native/blob/master/FAQ.md#q-how-do-i-enable-bitcode-for-my-kotlin-framework

https://github.com/JetBrains/kotlin-native/commit/6cb4d0c241f099367e47920c41a00c0cae72277f https://github.com/JetBrains/kotlin-native/commit/af769d958e0e277b0e81509d055c4a76416c49d3

**Comment by olonho** [comment]
Not exactly bitcode, what we support is adding markers showing Apple tools that produced code is aware about bitcode and could be used with it. So please try and let us know about your results.

**Comment by KaiJSchaefer** [comment]
Have someone any experience with this compiler option "-Xembed-bitcode"? Does apple reject apps with this marker ?

**Comment by webfrea-k** [comment]
We've just uploaded iOS app to iTunes Connect using -Xembed-bitcode without any problems. If your iOS project has ENABLE_BITCODE flag set, and any of the included frameworks does not provide bitcode, the archive operation will fail (and you cannot upload it to iTunes).

ps: iOS Framework that we are using was built from kotlin-native project with -xembed-bitcode marker.

**Comment by gimlids** [comment]
@webfrea-k you upload to AppStore connect. Was the app approved and did you publish it in the App Store?

**Comment by webfrea-k** [comment]
@gimlids no, we did not send it to approval. It was just for testing if technicality speaking app is compliant with AppStore. We were concerned if upload will fail due to some internal bitcode checking mechanism.

**Comment by amrox** [comment]
I'm confused - is bitcode working or not?

I did a quick and dirty test and it seemed to be working. I created a dummy Kotlin Native framework:

```
$(HOME)/lib/kotlin-native-macos-1.3.31/bin/kotlinc-native -produce framework -target ios_arm32  -Xembed-bitcode -o armv7/KLib lib.kt
$(HOME)/lib/kotlin-native-macos-1.3.31/bin/kotlinc-native -produce framework -target ios_arm64  -Xembed-bitcode -o arm64/KLib lib.kt
cp -R arm64/* universal
rm universal/KLib.framework/KLib
lipo armv7/KLib.framework/KLib arm64/KLib.framework/KLib -create -output universal/KLib.framework/KLib
```

Then I added the framework to a new app, archived it, and exported it with the "Rebuild from Bitcode" option using my enterprise cert. It seems to work fine.

<img width="536" alt="Screenshot 2019-06-15 07 58 32" src="https://user-images.githubusercontent.com/47537/59551186-95c7f080-8f43-11e9-94dd-7bac2cb874e0.png">

I've been searching for an alternative to C++ for sharing code between iOS and Android libraries, however lack of iOS bitcode support originally turned me away. If it is working, Kotlin/Native seems like an exciting option :)

**Comment by fluidsonic** [comment]
I've uploaded an iOS app to the App Store with the Bitcode option selected - there was no error or warning. Yet after processing there is no "Download dSYM" link. Doesn't that indicate that the binaries haven't been rebuilt?

**Comment by evant** [comment]
> Not exactly bitcode, what we support is adding markers showing Apple tools that produced code is aware about bitcode and could be used with it. So please try and let us know about your results.

So my understanding is iOS and tvOS would be supported, watchOS would not.

**Comment by SvyatoslavScherbina** [comment]
> Not exactly bitcode, what we support is adding markers showing Apple tools that produced code is aware about bitcode and could be used with it.

This will become false with 1.3.50, which is going to include real bitcode support.

> So my understanding is iOS and tvOS would be supported, watchOS would not.

Please note that "bitcode support" is not exactly the same as "watchOS support". The latter would also require other efforts.

**Comment by SvyatoslavScherbina** [comment]
Fixed in 1.3.50.
